pkg/server/api.go

```diff
@@ -186,31 +187,33 @@ func (sh *APIHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 type healthHandler struct{}
 
-// deriveSpecVersionFromUseragent returns a supported Ignition config spec version for a given Ignition user agent
-// or an error for unsuported user agents.
-// Only Ignition versions that come with support for either spec v2.2 or spec v3.1 are supported
-func deriveSpecVersionFromUseragent(useragent string) (*semver.Version, error) {
-	// for now, serve v2 if we receive a request without a useragent header with Ignition
-	// this happens if the user pings the endpoint directly (e.g. with curl)
-	// and we don't want to break existing behaviour
-	if !strings.HasPrefix(useragent, "Ignition/") {
+// deriveSpecVersionFromAcceptHeaderUseragent returns a supported Ignition config spec version for a given Accept header.
+// For non-Ignition Accept headers it defaults to config spec v2.2.0
+func deriveSpecVersionFromAcceptHeader(acceptHeader string) (*semver.Version, error) {
+	// serve v3 if we receive a request without an Ignition accept header.
+	// This happens if the user pings the endpoint directly (e.g. with curl)
+	// and we don't want to break existing behaviour.
+	// For Ignition v0.x, the accept header looks like:
+	// "application/vnd.coreos.ignition+json; version=2.4.0, application/vnd.coreos.ignition+json; version=1; q=0.5, */*; q=0.1".
+	// For v2.x, it looks like:
+	// "application/vnd.coreos.ignition+json;version=3.1.0, */*;q=0.1".
+	if !strings.HasPrefix(acceptHeader, "application/vnd.coreos.ignition+json") {
 		return semver.New("2.2.0"), nil
 	}
-	ignVersionString := strings.SplitAfter(useragent, "/")[1]
+	ignVersionString := strings.SplitAfter(acceptHeader, "=")[1]
+	ignVersionString = ignVersionString[0:5]
```

Comment:
Might be more robust to do something more like taking the whole string until we hit a `,`?  But eh.  Fine for now.